### PR TITLE
Fixes #58: Add FileReference type and use for FileShared and FilePublic

### LINKED ISSUE #58: FileShared and FilePublic contain reference to shared file
When the bot sees a file that has been uploaded he produces errors related to two messages. The messages are `file_shared` and `file_public`, and they now only reference the file by id. 

```
MessagePayload {messageId = 3, messageType = "message", messageChannel = Id {_getId = "C1E4KJ8M9"}, messageText = "<@U09EK8480|matthew> uploaded a file: <https://maul-t
ech.slack.com/files/matthew/F1E6AS0V8/dogs.jpeg|dogs.jpeg>"}
```

```
{"type":"file_shared","file_id":"F1E6AS0V8","user_id":"U09EK8480","file":{"id":"F1E6AS0V8"},"event_ts":"1465045015.197577"}
Error in $.file: failed to parse field file: key "timestamp" not present
Please report this failure to the github issue tracker
```

```
{"type":"file_public","file_id":"F1E6AS0V8","user_id":"U09EK8480","file":{"id":"F1E6AS0V8"},"event_ts":"1465045015.197578"}
Error in $.file: failed to parse field file: key "timestamp" not present
Please report this failure to the github issue tracker
```

### PR DESCRIPTION
**Pull request by matthewfranglen**
This is my attempt at fixing the issue. It may be terrible :smile:

### CODE REVIEW

**Comment by matthewfranglen** [comment]
I don't know why the travis test has failed. This works on my machine:

```
➜ SLACK_API_TOKEN=TOKEN stack test
slack-api-0.8: test (suite: tests)

rtm.start call successful
```

**Comment by mpickering** [comment]
I don't know why the tests fail so often on Travis. I don't think doing network IO is especially supported.

File: src/Web/Slack/Types/File.hs [hunk 1]
```diff
@@ -65,6 +67,7 @@ instance FromJSON Preview where
 
 data Preview = Preview { _previewText :: Text, _previewHighlight :: Text, _lines :: Int, _linesMore :: Int } deriving Show
 data FileUrl = URL { _private :: Text, _privateDownload :: Text } deriving Show
+data FileReference = FileReference { _id :: FileId } deriving Show
```
Comment (mpickering) [inline src/Web/Slack/Types/File.hs]: Please call this `_fileReferenceId` and then generate lenses as for the other datatypes (see examples on the lines below).

**Comment by mpickering** [comment]
If you could deal with the small comment then I will merge it.

**Comment by matthewfranglen** [comment]
I should be able to do that this afternoon.

**Comment by matthewfranglen** [comment]
Really it would be nice to have the file reference be a lazy file that only resolved if you looked at anything but the id.

**Comment by mpickering** [comment]
Yes it would. However the framework currently doesn't update the local state at all after the initialisation. This would be a very nice feature to have but not something I will ever implement myself as I don't use the library.

**Comment by bw-matthew** [comment]
I have updated the PR btw.

**Pull request merged**